File: backend/app/api/molds.py
"""금형 대시보드 조회 API.

`Depends(get_app_state)` 를 쓰지 않는다 — 데이터가 코드 안의 샘플이라 settings/DB
가 필요 없다. 서브프로젝트 ④(영속화)에서 DB 가 필요해지면 그때 의존성을 추가한다.
"""
from fastapi import APIRouter, HTTPException

from app.molds.query import filter_options, get_mold, list_molds
from app.molds.schemas import FilterOptions, MoldDetail, MoldStatus, MoldSummary

router = APIRouter()


# 주의: 이 라우트는 /molds/{mold_no} 보다 **먼저** 선언해야 한다. FastAPI 는
# 선언 순서대로 매칭하므로, 순서가 뒤바뀌면 "filters" 가 금형 번호로 잡혀
# 404 가 난다.
@router.get("/molds/filters", response_model=FilterOptions)
async def get_filter_options() -> FilterOptions:
    return filter_options()


@router.get("/molds", response_model=list[MoldSummary])
async def get_molds(
    status: MoldStatus | None = None,
    line: str | None = None,
    machine: str | None = None,
    q: str | None = None,
) -> list[MoldSummary]:
    return list_molds(status=status, line=line, machine=machine, q=q)


@router.get("/molds/{mold_no}", response_model=MoldDetail)
async def get_mold_detail(mold_no: str) -> MoldDetail:
    mold = get_mold(mold_no)
    if mold is None:
        raise HTTPException(status_code=404, detail="mold not found")
    return mold

File: frontend/src/api/molds.ts
import type { FilterOptions, MoldDetail, MoldFilters, MoldSummary } from "../types/mold";
import { API_HEADERS } from "./headers";

const BASE = "/api";

async function json<T>(res: Response): Promise<T> {
  if (!res.ok) throw new Error(`HTTP ${res.status}`);
  return res.json() as Promise<T>;
}

/** 값이 설정된 필터만 쿼리스트링에 담는다. 빈 값을 보내면 서버가 ""로 필터해 0건이 된다. */
function toQuery(filters: MoldFilters): string {
  const params = new URLSearchParams();
  const q = filters.q.trim();
  if (q) params.set("q", q);
  if (filters.status !== "all") {
    params.set("status", filters.status);
    // 라인/호기는 상태에 종속이다. 상태가 '전체'면 이 값들은 의미가 없으므로
    // 보내지 않는다 — 화면에서 드롭다운이 사라져 있어도 스토어에 옛 값이
    // 남아 있을 수 있어, 여기서 한 번 더 막는다.
    if (filters.line) params.set("line", filters.line);
    if (filters.machine) params.set("machine", filters.machine);
  }
  const qs = params.toString();
  return qs ? `?${qs}` : "";
}

export function listMolds(filters: MoldFilters): Promise<MoldSummary[]> {
  return fetch(`${BASE}/molds${toQuery(filters)}`, {
    method: "GET",
    headers: API_HEADERS,
  }).then(json<MoldSummary[]>);
}

export function getMold(moldNo: string): Promise<MoldDetail> {
  return fetch(`${BASE}/molds/${encodeURIComponent(moldNo)}`, {
    method: "GET",
    headers: API_HEADERS,
  }).then(json<MoldDetail>);
}

export function getFilterOptions(): Promise<FilterOptions> {
  return fetch(`${BASE}/molds/filters`, {
    method: "GET",
    headers: API_HEADERS,
  }).then(json<FilterOptions>);
}
